### PR TITLE
cmake: prework towards supporting board groups

### PR DESCRIPTION
**Pull request by galak**
These set of changes cleanup how we determine the ARCH dir based on the path to the board dir.  This allows us to support additional directories between boards/ARCH/ and BOARD_DIR.  So we can have:

boards/arm/96_boards/nitrogen
boards/arm/96_boards/neonkey
boards/arm/bbc_microbit

### CODE REVIEW

**Review by carlescufi — APPROVED** [review]
LGTM

**Review by nashif — DISMISSED** [review]
This PR does not introduce grouping, just prepares for them, so the title is confusing. If the intent is to support grouping, we need more context and also an example of how this would work and addressing any issue that might come with grouping, for example
- out of tree
- documentation
- why do we need grouping in the first place and how this would work when we share headers or DTS etc.

**Comment by galak** [comment]
> This PR does not introduce grouping, just prepares for them, so the title is confusing. If the intent is to support grouping, we need more context and also an example of how this would work and addressing any issue that might come with grouping, for example
> 
> * out of tree
> * documentation
> * why do we need grouping in the first place and how this would work when we share headers or DTS etc.

This was intended to provide the infrastructure or cleanup towards supporting it for #12195.  I think the cleanups/changes the first to commits make are reasonable changes to go in now.  I don't really have the bandwidth or interest in trying to propose a fuller solution for grouping.

Is that ok?

**Comment by carlescufi** [comment]
@nashif 

>out of tree

Is there an issue there? The old layout is supported as before.

>documentation

Sphinx just looks for the `.rst` files which will happen to be one level down, I don't think this will change the documentation, though we could certainly benefit from grouping in the doc.

>why do we need grouping in the first place and how this would work when we share headers or DTS etc.

I think the only motivation right now is @pabigot's requirement to share DT files among similar boards (the Particle series).

**Comment by nashif** [comment]
@carles I do not know of any issues, however, if we are going to add one more level to the boards directory, we need a proposal and a discussion and we need to make sure it does not break anyone. How we find ARCH and _defconfig files and how we discover boards in general has it share of ad-hoc solutions that we really need to think about it a bit more before we decide how to support grouping of boards. 

This PR does not really do anything yet, i know and probably it can be merged, however I am opposed to adding another level in the board directory without a thorough discussion and identifying alternatives.

I have just seen a proposal to do that in a consistent way using soc name across the board for example, see 

https://github.com/zephyrproject-rtos/zephyr/pull/12108#issuecomment-447884050

How is this going to work in general, i.e. how do you share information (headers) among boards in the same family? Who is going to move all boards into this structure?

**Comment by pabigot** [comment]
Currently in #12195 I do this in the board-specific DTS files to access the shared information:
```
#include <vendor/particle/mesh_feather.dtsi>
```
where `vendor` here is subordinate to `${ZEPHYR_BASE}/dts` which is in the include path provided by `dts.cmake`.

Exactly how the changes in this PR are intended to be used isn't clear.  Is it like this:

Assuming a directory layout:
```
boards/arm/particle/dts
boards/arm/particle/particle_argon
boards/arm/particle/particle_boron
boards/arm/particle/particle_xenon
```
change the include to:
```
#include <particle/dts/mesh_feather.dtsi>
```
assuming this change to `dts.cmake` is made:
```
diff --git a/cmake/dts.cmake b/cmake/dts.cmake
index 073c52f7fd..39f3d2a253 100644
--- a/cmake/dts.cmake
+++ b/cmake/dts.cmake
@@ -96,6 +96,7 @@ if(CONFIG_HAS_DTS)
     -x assembler-with-cpp
     -nostdinc
     -isystem ${DTS_APP_INCLUDE}
+    -isystem ${ZEPHYR_BASE}/boards/${ARCH}
     -isystem ${ZEPHYR_BASE}/include
     -isystem ${ZEPHYR_BASE}/dts/${ARCH}
     -isystem ${ZEPHYR_BASE}/dts
```
?  This seems to work, but the `dts.cmake` change should probably be part of this PR.

**Comment by pabigot** [comment]
The board documentation appears to generate correctly when placed within an intervening family directory.

File: cmake/app/boilerplate.cmake [hunk 1]
```diff
@@ -219,8 +219,16 @@ if(NOT BOARD_DIR)
...
+# Determine arch by the directory name after 'boards'
+# We search until we find 'boards' so we can have any number
+get_filename_component(PARENT_DIR ${BOARD_DIR} PATH)
+get_filename_component(DIR_NAME  ${PARENT_DIR} NAME)
+while(NOT ("boards" STREQUAL ${DIR_NAME}))
+  set(ARCH ${DIR_NAME})
+  get_filename_component(GPARENT_DIR ${PARENT_DIR} PATH)
+  get_filename_component(DIR_NAME   ${GPARENT_DIR} NAME)
+  set(PARENT_DIR ${GPARENT_DIR})
+endwhile(NOT ("boards" STREQUAL ${DIR_NAME}))
```
Comment (SebastianBoe) [inline cmake/app/boilerplate.cmake]: Using 'while' and saying "We search until we find 'boards' so we can have any number" implies that we support arbitrary nesting, but we don't, we support one level of nesting as this commit demonstrates:

https://github.com/zephyrproject-rtos/zephyr/pull/12377/commits/b34cf2baf5314da83047a003ba6f8d0102dca1bc

The code and comments should reflect this.

**Review by SebastianBoe — CHANGES_REQUESTED** [review]

File: cmake/app/boilerplate.cmake (hunk 1 above)
Comment (galak) [reply cmake/app/boilerplate.cmake]: I can change the commit message, but I want the code to stay the same, such that regardless of what struct of dir, we determine ARCH based on looking for 'boards'.

**Review by pabigot — CHANGES_REQUESTED** [review]
Needs this patch, then I think it handles the case for board families required for #12195.
```
diff --git a/cmake/dts.cmake b/cmake/dts.cmake
index 073c52f7fd..39f3d2a253 100644
--- a/cmake/dts.cmake
+++ b/cmake/dts.cmake
@@ -96,6 +96,7 @@ if(CONFIG_HAS_DTS)
     -x assembler-with-cpp
     -nostdinc
     -isystem ${DTS_APP_INCLUDE}
+    -isystem ${ZEPHYR_BASE}/boards/${ARCH}
     -isystem ${ZEPHYR_BASE}/include
     -isystem ${ZEPHYR_BASE}/dts/${ARCH}
     -isystem ${ZEPHYR_BASE}/dts
```

If documentation of how to lay out the family directory is yet to be added before this can be merged I would like it to support this:
```
boards/arm/particle/dts
boards/arm/particle/particle_argon
boards/arm/particle/particle_boron
boards/arm/particle/particle_xenon
```

**Pull request closed without merging**